src/io/aleph/dirigiste/Pool.java

```diff
@@ -406,29 +436,37 @@ public void run() {
     public void acquire(final K key, final AcquireCallback<V> callback) {
         final long start = System.nanoTime();
 
-        Queue q = queue(key);
-        AcquireCallback<V> wrapper =
-            new AcquireCallback<V>() {
-                    public void handleObject(V obj) {
+        try {
+            // To prevent the queue from being deleted by the startControlLoop method (which runs on another thread) as
+            // soon as it has been created, we need to mark the Queue as in use after acquired the exclusive lock.
+            _lock.lock();
+            _lockedQueues.compute(key, (__, v) -> v == null ? 1 : v + 1);
+            _lock.unlock();
 
+            Queue q = queue(key);
+            AcquireCallback<V> wrapper =
+                    obj -> {
                         // do all the latency bookkeeping
                         long acquire = System.nanoTime();
                         _queueLatencies.sample(key, acquire - start);
                         _start.put(obj, start);
 
                         callback.handleObject(obj);
-                    }
-            };
-        boolean success = q.take(wrapper, false);
-
-        // if we didn't immediately get an object, try to create one
-        if (!success) {
-            try {
-                addObject(key);
-            } catch (Throwable e) {
-                q.cancelTake(wrapper);
-                throw new RuntimeException(e);
+                    };
+            boolean success = q.take(wrapper, false);
+
+            // if we didn't immediately get an object, try to create one
+            if (!success) {
+                try {
+                    addObject(key);
+                } catch (Throwable e) {
+                    q.cancelTake(wrapper);
+                    throw new RuntimeException(e);
+                }
             }
+        } finally {
+            // This call is safe, there is no reason the key wont be present on the locked queues at this point.
+            _lockedQueues.compute(key, (__, v) -> v - 1);
         }
```

Comment:
This is just for initialization purposes, right? The `_lockedQueues` key value could be set to 0 when `acquire` ends, even though something is still holding the object. This is _solely_ to prevent `adjust` from removing it until we've added at least one object to the new queue, right?

Reply:
Using a KeySet instead of a ConcurrentHashMap directly is even better actually: 391325b

Reply:
Works for me

Reply:
I will add more tests on a concurrent environement to ensure it's working as expected.
After a second thought, maybe tracking the number of pending `acquire` was necessary to prevent
dropping the queue while trying to acquire an element.

`acquire` => `release` => `acquire`

Reply:
I reverted back the code to what I proposed initially.
It seems safer to me and I can see situations where we could remove the queue
while trying to create an object on it.